Numerous improvements and refactorings

chrisspen opened the pull request:
I've made several significant changes:
1. I've added support for serving files from the local file system. When a file is uploaded, a database record is still created but the file is copied to the local file system where it is served from. Its hash is recorded, in a *.hash file on the file system, as well as in the database in a new `content_hash` field. The database serves as the master repository, so if the file system's hash differs (e.g. because a user uploaded a new file from a different server), it's replaced with the version from the database. There are four management commands to aid in various maintenance actions, such as loading files into the database, dumping everything out of the database, deleting orphaned files and regenerating file hashes.
2. I've refactored the general package layout and setup.py to make it a little easier to manage and test. Namely, all unittests can be run directly from setup.py with `python setup.py test`.
3. I've added Python3 support. This can be tested by running `python setup.py test --pv=3`

chrisspen committed:
Modified database storage to fallback to file storage. Added management
commands to handle bulk load and dump of files to/from the filesystem.

chrisspen committed:
Added data migration to auto-load existing files into the database.
Updated documentation to note differences from parent.

chrisspen committed:
Fixed typo.

chrisspen committed:
Added management command to handle bulk file deletions.

chrisspen committed:
Updated version number.

chrisspen committed:
Fixed bug in dump management command.

chrisspen committed:
Updated version.

chrisspen committed:
Added option to load command to allow limiting file load by app model.

Commit:
Limited debugging output.

Commit:
Fixed unittest command.

Commit:
Added feature to define custom user and group to set ownership with when
writing files.

Commit:
Updated version number.

Commit:
Improved performance of database_files_dump by only exporting
content from the database if there's a difference in file hashes.

Commit:
Added a dryrun option to the cleanup command. Fixed minor bug in content
read caused by partially pre-read file object.

Commit:
Added a unittest for hashing functions.

Commit:
Fixing hashing functions to be consistent in Python2.6. Added management
command to regenerate hashes.

Commit:
Fixed management command.

Commit:
Added file hash caching on the filesystem.

Commit:
Updated version.

Commit:
Modified the default behavior to automatically export files from the
database if they don't exist on the filesystem.

Commit:
Updated version number.

Commit:
Modified file dump command to update hash upon write and allow targeting
of specific files.

Commit:
Added configurable file URL constructor.

Commit:
Updated version.

Commit:
Improved performance of cleanup command.

Commit:
Added view to automatically handle retrieving files from either the
filesystem or database.

Commit:
Updated version.

Commit:
Fixed bug causing file to be unnecessarily re-written on every request.

Commit:
Added table name.

Commit:
Updated ignore.

Commit:
Added test image Python markup.

Commit:
Added test image.

Commit:
Added support for Python3.

Commit:
Updated version.

chrisspen commented:
Let me know if you have any questions. I've tried to document as much as I can, but I haven't had time to get to everything. However, unittests should cover the bulk of the new functionality.

Btw, If you don't want to maintain the project any longer, I'm willing to take over.

chrisspen commented:
Any update on this?

chrisspen commented:
I'll interpret the lack of response to mean you're not interested. I'll publish my changes as a separate package.

Commit:
Repackaged. Fixed Python3 unittest due to South bug. Updated docs.

bfirsh commented:
Sorry @chrisspen – I am definitely interested, but haven't got time to review right now. I'll leave this open for when I have time to take a look. :)

Commit:
Broadened exception handling for pypandoc.

Commit:
Added error handling for cases where path is passed.

Commit:
Fixed error handling bug in setup.

Commit:
Added option to specify filename to cleanup.

richardxia committed:
Move South migrations to south_migrations.

richardxia committed:
Add Django-1.7-style migrations.

chrisspen committed:
Merge pull request #1 from richardxia/add_django_1.7_support

Add Django 1.7 support

richardxia committed:
Add south_migrations to distribution.

chrisspen committed:
Merge pull request #2 from richardxia/fix_setup_packages

Add south_migrations to package.

rhunwicks committed:
Make Django 1.7 migrations compatible with Python3

rhunwicks committed:
Update README.md to include instructions for changing urls.py

chrisspen committed:
Merge pull request #3 from rhunwicks/master

Make Django 1.7 migrations compatible with Python3

chrisspen committed:
Added support for Django<=1.10.1.

chrisspen committed:
Merge branch 'master' of
git+ssh://github.com/chrisspen/django-database-files-3000

# Conflicts:
#	database_files/migrations/0001_initial.py

chrisspen committed:
Updated docs.

Commit:
the bytestring was breaking the migration.

If there is no particular reason to have a bytestring in place of the original package's string db column, then we can revert to the string itself.

chrisspen committed:
Merge pull request #4 from niranjanthilak/patch-1

the bytestring was breaking the migration.

chrisspen committed:
changed travis link

chrisspen committed:
fixed unittests for django>=1.7

chrisspen committed:
Fixed unittests for all supported Python+Django versions.

chrisspen committed:
Fixed code formatting.

chrisspen committed:
Fixed undefined variable.

chrisspen committed:
Cleaned up setup.

chrisspen committed:
Updated shields.

samorajp committed:
Typo fix in README.

Typo fix in README.

chrisspen committed:
Merge pull request #5 from samorajp/patch-1

Typo fix in README.

pyup-bot committed:
Pin south to latest version 1.0.2

chrisspen committed:
Merge pull request #6 from chrisspen/pyup-initial-update

Initial Update

YamaoMaoen committed:
Drop Django 1.6 support

chrisspen committed:
Merge pull request #7 from YamaoMaoen/master

Drop Django 1.6 support

chrisspen committed:
Extended django version. Updated ignore

chrisspen committed:
Merge branch 'master' of git+ssh://github.com/chrisspen/django-database-files-3000

chrisspen committed:
Updated travis config to include python 3.6

chrisspen committed:
Updated travis config to include python 3.6

chrisspen committed:
Dropped Python 3.3 support.

chrisspen committed:
Updated version

chrisspen committed:
Refactored default settings initialization to work with Django 1.11.

JordanReiter committed:
trusty does not have 3.5 distribution

Switching to xenial as recommended here: https://travis-ci.community/t/unable-to-download-python-3-7-archive-on-travis-ci/639

JordanReiter committed:
Python 3.4 is EOL & no longer available & pylint > 2.0 does not support 2.7

Commit:
return None to fix inconsistent-return-statements warning for pylint

chrisspen committed:
Merge pull request #24 from JordanReiter/patch-1

trusty does not have 3.5 distribution

JordanReiter committed:
Import location changed in Django 2.0+

JordanReiter committed:
Import location changed in Django 2.0+

chrisspen committed:
Merge pull request #27 from JordanReiter/patch-2

Import location changed in Django 2.0+

chrisspen committed:
Merge pull request #28 from JordanReiter/patch-2

Import location changed in Django 2.0+

Commit:
Django requirement

Commit:
convert bytes to str before saving to TextField

Commit:
Upgraded tests

chrisspen committed:
Merge pull request #29 from JordanReiter/patch-2

Django requirement

chrisspen committed:
Increased version.

chrisspen committed:
Added yapf config.

chrisspen committed:
Updated readme attachment in setup.py

chrisspen committed:
Removed Python2.7 support from test suite.

chrisspen committed:
Removed Python2.7 support from test suite.

chrisspen committed:
Added test script.

chrisspen committed:
Removed Python2.7 support from test suite.

chrisspen committed:
Removed Python2.7 support from test suite.

chrisspen committed:
Removed Python2.7 support from test suite.

chrisspen committed:
Removed Python2.7 support from test suite.

chrisspen committed:
Forced name methods to force string.
